Restrict access to boards

kingraphaii opened the pull request:
Related issues: [please specify]

## Description:

restricted access to user boards

## Screenshots/videos

<!-- If there is a visual component to what you did, please save us time by adding a screenshot. You can even link to a video demonstrating your feature, that would be cool too -->

## I solemnly swear that:

- [x] My code follows the style guidelines of this project
- [x] I have merged the develop branch into my branch and fixed any merge conflicts
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have tested new or existing tests and made sure that they pass

kingraphaii committed:
Refactor user board and user board column views

sheenarbw commented on backend/frontend/views.py:
```diff
@@ -50,40 +54,93 @@
 
-@user_passes_test(is_super)
-def user_board(request, user_id):
+def user_has_view_access(user, request):
```
This is not super clear...
is `user` the logged in user? Or is it the user being viewed?

sheenarbw replied:
And it looks like the only reason we are passing in request is so that we can access request.user. Why not just pass in the logged in user as an instance? 

eg: `user_has_view_access(viewed_user,logged_in_user)` or something like that?

sheenarbw commented on backend/frontend/views.py:
```diff
@@ -50,40 +54,93 @@
+    is_trusted_reviewer = [
+        request.user.has_perm(Team.PERMISSION_TRUSTED_REVIEWER, team)
+        for team in user_teams
+    ]
```
This code seems quite inefficient. If a person has "manage cards" permission on a team then it would make sense to return immediately. Instead, we do many database queries before returning.

This function is going to be called really often, it needs to be quick!

If the viewed user is in 3 teams then this is 3x5 = 15 distinct db queries for every single permission check.

Read up on django guardian and see if there is a more efficient way to do it. And if you have to make so many queries in the worst case scenario then please just return as early as possible.

sheenarbw commented on backend/frontend/views.py:
```diff
@@ -50,40 +54,93 @@
+def user_has_view_access(user, request):
+    user_teams = user.teams()
+    can_manage_cards = [
+        request.user.has_perm(Team.PERMISSION_MANAGE_CARDS, team) for team in user_teams
```
This is not DRY, why are you typing out all the view permissions one by one? They all exist in a data structure that can be imported and looped over. There is a good chance that we will be adding new permissions to that data structure in the near future, when we do that then this code will bug out unless someone remembers to come and update it. It's better to be DRY

sheenarbw commented on backend/frontend/views.py:
```diff
@@ -50,40 +54,93 @@
+    return has_view_access or user.id == request.user.id or request.user.is_superuser
+
+
+class UserBoard(LoginRequiredMixin, UserPassesTestMixin, TemplateView):
```
This code is much longer and more complicated than the function-based version. What are the benefits of this approach versus just using a user_passes_test decorator? 

My understanding is that the only real reason to make use of class based views is to make use of inheritance and it doesn't look like we are making much use of that - we aren't putting our own functionality into a base class and inheriting from that. 

If there is a legit reason to use class based views then please say what that reason is. But generally we should try to keep things simple as much as possible.

kingraphaii replied:
couldn't pass an extra argument to `user_passes_test` and also request is not accessible in there either, so I thought the CB version could do it without much deliberation. So I was thinking either this or putting the checks inside each view which wouldn't look good. Would creating a custom `user_passes_test` decorator that takes in the request object as well be a good implementation of this?

sheenarbw replied:
Putting the checks inside each view wouldn't do.

I made a middleware that lets you get the currrent request. 

```
from threadlocal_middleware import get_current_request
request = get_current_request()
...
```

Does that help?  Or do you still think classes are better?

kingraphaii replied:
Noice! This should do, getting the request was the only issue I had. Thanks!

sheenarbw commented on backend/frontend/views.py:
```diff
@@ -50,40 +54,93 @@
+        or any(can_review_cards)
+        or any(is_trusted_reviewer)
+    )
+
```
The overall algorithm should be:

```
# do the cheap checks first. Doing this wont hit the db at all so get it out of the way

If user.id == request.user.id or request.user.is_superuser:
    return True 

# now do the expensive checks if needed 
```

sheenarbw reviewed (changes requested):
This needs a bit of work. It needs to be efficient and DRY. And I suspect it can be done a lot more simply

kingraphaii committed:
Simplify auth checks for board access

sheenarbw commented on backend/frontend/views.py:
```diff
@@ -50,22 +57,43 @@
 
 
-@user_passes_test(is_super)
+def has_view_access(logged_user):
```
logged_user? You mean logged_in_user?

sheenarbw replied:
Also, sometimes a person can have view access over a whole team. This function name can be more explicit.

Maybe `can_view_user` or something like that? Then we can do another one for `can_view_team`. Or is the intention of this function to be used on user views and team views?

sheenarbw replied:
I did originally put the is_super function in views.py, but it was because it is a 2 line function we will probably throw away. This is getting a bit long.

Can you think of a better way to organize the code? 

I also think it would be good to write tests.

sheenarbw reviewed (changes requested):
Much better. Now to make sure the code is well organised and tested.

kingraphaii committed:
Refactor view access control in user_board views

kingraphaii committed:
Merge branch 'develop' into restricted-access-to-cards

kingraphaii committed:
front-end tests for controlled board access

Ngoakor12 commented on backend/frontend/tests/test_restricted_user_board_access.py:
```diff
@@ -0,0 +1,156 @@
+
+        self.viewed_user_team = TeamFactory()
+        self.viewed_user = UserFactory(
+            email="learner.umuzi.org", is_staff=True  # TODO: remove this
```
A forgotten todo? If it should be there please leave a reason or more info just like login todo on line 117.

kingraphaii replied:
Not really, more of a reminder. After we merge the proper login urls we wont need is_staff=True. This is sort of a workaround for getting a valid session via the admin portal

Ngoakor12 commented on backend/frontend/tests/test_restricted_user_board_access.py:
```diff
@@ -0,0 +1,156 @@
+        self.viewed_user_team.user_set.add(self.viewed_user)
+
+        self.user_without_access = UserFactory(
+            email="user_without_access.umuzi.org", is_staff=True  # TODO: remove this)
```
same here

kingraphaii committed:
Merge branch 'develop' into restricted-access-to-cards

kingraphaii committed:
Remove unnecessary is_staff=True instances and add explanatory comment

Ngoakor12 reviewed (approved):
There are some merge conflicts, but the changes look addressed.

kingraphaii committed:
Merge branch 'develop' into restricted-access-to-cards

kingraphaii committed:
Update login page and fix test cases

sheenarbw commented on backend/frontend/views.py:
```diff
@@ -69,6 +71,28 @@ def is_super(user):
+    if len(viewed_user_teams):
+        checker = ObjectPermissionChecker(logged_in_user)
+        checker.prefetch_perms(viewed_user_teams)
+        for view_permission in Team.PERMISSION_VIEW:
```
This can still be more efficient. Here's something to try out:

```
def work():
     answer = random.choice([True,False])
     print(f"work - {answer}")
     return answer

# what is the difference between the following:
any([work() for _ in range(10)])
any((work() for _ in range(10)))
```

One is a list comprehension, and the other one uses a tuple. The second one is more like a generator. It wont do ALL the work up front.

Now imagine if `work` was an expensive function, eg, what if it queried a database..

TLDR; a tuple is a way better choice here. Not just slightly better.

sheenarbw replied:
Also, I'm not sure, can has_perm accept a list? Like `checker.has_perm(Team.PERMISSION_VIEW, team)`. If it can then we should definately not loop.  You've probably already checked this, but figured I would mention it just in case.

sheenarbw commented on backend/frontend/tests/test_restricted_user_board_access.py:
```diff
@@ -0,0 +1,159 @@
+from .frontend_test_mixin import FrontendTestMixin
+
+
+class TestUserBoardControlledAccess(FrontendTestMixin):
```
A better name would be something like TestUserBoard Permissions or ...Authorization.

Controllers are a whole thing.

sheenarbw commented on backend/frontend/tests/test_restricted_user_board_access.py:
```diff
@@ -0,0 +1,159 @@
+        login_url = "/login"
+        final_page_url = self.page.url
+
+        self.assertIn(login_url, final_page_url)
```
What if the user is already logged in? Should something else happen? Eg, maybe they should get a page like "you do not have permission to view this page"

sheenarbw reviewed (changes requested):
There is one UX thing that seems a bit off, and an efficiency thing to do with how Python deals with tuples versus lists.

kingraphaii committed:
Refactor user board access control and views

This commit refactors the user board access control and views in the frontend. The `TestUserBoardControlledAccess` class has been renamed to `TestUserBoardAuthorization` for clarity. The `user_passes_test_or_forbidden` decorator has been added to the views to check if the user has permission to access the page and return a 403 Forbidden response if not. This ensures that unauthorized users cannot view other users' boards.

kingraphaii committed:
Merge branch 'develop' into restricted-access-to-cards

Ngoakor12 commented:
Merge conflicts, otherwise, changes look addressed.

kingraphaii committed:
Merge branch 'develop' into restricted-access-to-cards

sheenarbw commented on backend/frontend/views.py:
```diff
@@ -72,6 +76,51 @@ def is_super(user):
+            if test_func(request.user):
+                return view_func(request, *args, **kwargs)
+
+            return HttpResponseForbidden(
```
If the user is logged in then they should get a forbidden message. If they are not logged in they should be directed to the login page.
